In [2]:
import pandas as pd
import numpy as np
import scipy.stats as stats
from scipy.stats import shapiro

In [11]:
df=[8800, 9500, 9600, 10500, 10400, 9600, 10300, 9500, 9600, 7100, 10100, 7200]
df

[8800, 9500, 9600, 10500, 10400, 9600, 10300, 9500, 9600, 7100, 10100, 7200]

- H0 : 진공관 수명은 1만시간이다
- H1 : 진공관 수명은 1만시간이 아니다.

In [4]:
shapiro(df)

ShapiroResult(statistic=0.8105747699737549, pvalue=0.012388509698212147)

귀무가설을 기각한다. 즉 정규성을 따르지 않는다.

In [5]:
mean0=10000
stats.wilcoxon(df-mean0, alternative='two-sided')

WilcoxonResult(statistic=15.5, pvalue=0.06396484375)

귀무가설을 기각할 수 없다. 즉 진공관 수명은 1만시간이다.

### 문제2

In [7]:
median=np.median(df)
median

9600.0

In [9]:
from collections import Counter
Counter(df)

Counter({8800: 1,
         9500: 2,
         9600: 3,
         10500: 1,
         10400: 1,
         10300: 1,
         7100: 1,
         10100: 1,
         7200: 1})

9600이 세개 있으므로 제외해줌

In [12]:
while 9600 in df:
    df.remove(9600)

In [14]:
df=np.array(df)

In [15]:
mean0=10000
df
stats.wilcoxon(df-mean0, alternative='two-sided')

WilcoxonResult(statistic=11.0, pvalue=0.203125)

In [25]:
mean0=60
df=pd.DataFrame({'num':[52,50,62,75,26,45,62,35,57,14]})
df

num
0   52
1   50
2   62
3   75
4   26
5   45
6   62
7   35
8   57
9   14

In [33]:
median=df.num.median()

In [34]:
median

51.0

In [35]:
df=df[df['num']!=51]

In [36]:
shapiro(df)

ShapiroResult(statistic=0.965764582157135, pvalue=0.849041759967804)

귀무가설을 기각할 수 없다. 즉 정규분포를 따른다

In [38]:
mean0=60
stats.wilcoxon(df.num-mean0, alternative='two-sided')

WilcoxonResult(statistic=9.5, pvalue=0.064453125)

검정통계량은 9.5이고 pvalue가 0.05보다 작으므로 귀무가설을 기각할 수 없다. 즉 모평균은 60이다.

In [43]:
mean0=60
s_mean=np.mean(df['num'])
u_var=df['num'].var(ddof=1)
n=len(df['num'])
rv=stats.t(df=n-1)

In [44]:
t=(s_mean-mean0)/np.sqrt(u_var/n)
t

-2.099292080780179

In [45]:
p=rv.cdf(t)*2

In [46]:
p

0.06519313277276138

In [47]:
alpha=0.05
rv.interval(1-alpha)

(-2.2621571627409915, 2.2621571627409915)

In [49]:
lcl=s_mean-rv.isf(0.025)*np.sqrt(u_var/n)
ucl=s_mean-rv.isf(0.975)*np.sqrt(u_var/n)
lcl, ucl

(34.65351236347088, 60.94648763652911)

# 문제

In [60]:
df=pd.DataFrame({'index':["1.5-2.5", "2.5-3.5", "3.5-4.5"], "사회과학":[15,60,24], "자연과학":[25,69,6], "공학":[10, 77,14]})
df.set_index("index", inplace=True)
df

사회과학  자연과학  공학
index                  
1.5-2.5    15    25  10
2.5-3.5    60    69  77
3.5-4.5    24     6  14

In [54]:
from scipy.stats import chi2_contingency
chi, p, df, expect=chi2_contingency(df)
chi, p

(20.213713780910304, 0.0004531468852769105)

검정통계량은 20.21이고 p 값은 0.05보다 작으므로 귀무가설을 기각한다. 즉 성적과 과목은 독립이 아니다.

In [55]:
expect

array([[16.5       , 16.66666667, 16.83333333],
       [67.98      , 68.66666667, 69.35333333],
       [14.52      , 14.66666667, 14.81333333]])

기댓값은 다음과 같다.

In [63]:
사회과학, 자연과학, 공학=df.sum()
사회과학, 자연과학,  공학

A, B, C=data.sum(axis=1)
n=사회과학+자연과학+공학
n

300

In [64]:
ad_ef=pd.DataFrame({'사회과학':[A*사회과학/n, B*사회과학/n, C*사회과학/n],'자연과학':[A*자연과학/n, B*자연과학/n, C*자연과학/n], '공학':[A*공학/n, B*공학/n, C*공학/n]}, index=df.index)
ad_ef

사회과학       자연과학         공학
index                               
1.5-2.5  16.50  16.666667  16.833333
2.5-3.5  67.98  68.666667  69.353333
3.5-4.5  14.52  14.666667  14.813333

In [65]:
y=((df-ad_ef)**2/ad_ef).sum().sum()
y

20.213713780910307

In [67]:
rv=stats.chi2(4)
1-rv.cdf(y)

0.0004531468852768672

# 문제

In [70]:
df=pd.DataFrame({'num':[3.1, 3.3, 3.5, 3.7, 3.9, 4.1, 4.3, 4.4, 4.7]})
n=9

In [74]:
rv=stats.t(df=n-1)
alpha=0.05
s_mean=np.mean(df.num)
p_var=np.var(df.num, ddof=1)
lcl=s_mean-2.31*np.sqrt(p_var/n)
ucl=s_mean+2.31*np.sqrt(p_var/n)

In [75]:
lcl, ucl

(3.4770208544679386, 4.300756923309841)

모 표준편차=0.05, 모분산=0.05**2

In [76]:
rv=stats.norm()
p_var=0.04**2
lcl=s_mean-1.96*np.sqrt(p_var/n)
ucl=s_mean+1.96*np.sqrt(p_var/n)
lcl, ucl

(3.8627555555555566, 3.915022222222223)

In [78]:
result=0.01*(370/(370+15))/(0.01*370/(370+15)+0.99*10/(10+690))
result

0.4045926735921268

# 문제

In [79]:
mean_A=5.7
mean_B=5.6
var_A=0.03**2
var_B=0.04**2
n_A=12
n_B=12

#A, B 두 생산라인의 차이는 없다.
#A, B 두 생산라인 사이의 차이는 있다.

In [80]:
sp=np.sqrt(((n_A-1)*var_A+(n_B-1)*var_B)/(n_A+n_B-2))
sp

0.035355339059327376

In [81]:
t=(mean_A-mean_B)/(sp*np.sqrt(1/n_A+1/n_B))
t

6.928203230275546

In [82]:
rv=stats.t(df=n_A+n_B-2)
p=(1-rv.cdf(t))*2

In [83]:
p

5.892973662380285e-07

귀무가설을 기각한다. 즉 두 생산라인 사이의 차이는 있다.

# 문제

alpha=0.1, 추정오차한계(d)=0.05

In [85]:
import scipy.stats as stats
from scipy.stats import shapiro
rv=stats.norm()
alpha=0.1
z=rv.isf(alpha/2)
d=0.05
n=1/4*(z/d)**2
n
round(n,0)

271.0

In [86]:
z

1.6448536269514729

# 문제

d=10, 모표준편차=30, 신뢰수준 alpha0.1

In [92]:
alpha=0.1
sigma=30
rv=stats.norm()
z=rv.isf(alpha/2)
d=10
n=(z*sigma/d)**2
round(n,1)

24.3

n=25개

# 문제

d=3, alpha=0.05, sigma=np.sqrt(64)

In [93]:
alpha=0.05
sigma=np.sqrt(64)
rv=stats.norm()
z=rv.isf(alpha/2)
d=3
n=(z*sigma/d)**2

In [94]:
round(n,1)

27.3

In [95]:
n=28

# 문제

d=0.03, alpha=0.05, p=0.35

In [96]:
alpha=0.05
p=0.35
rv=stats.norm()
z=rv.isf(alpha/2)
d=0.03
q=1-0.35

n=p*q*(z/d)**2
n

971.0354241199045

n=972

# 문제

In [97]:
p=8/400
n=400
alpha=0.05
q=1-p
rv=stats.norm()
rv.isf(alpha/2)
lcl=p-rv.isf(alpha/2)*np.sqrt(p*q/n)
ucl=p-rv.isf(1-alpha/2)*np.sqrt(p*q/n)

lcl, ucl

(0.006280252108219619, 0.03371974789178038)

In [99]:
p1=291/300
p2=285/300
q1=1-p1
q2=1-p2
n=300

alpha=0.05
rv=stats.norm()
lcl=(p1-p2)-rv.isf(alpha/2)*np.sqrt(p1*q1/n+p2*q2/n)
ucl=(p1-p2)-rv.isf(1-alpha/2)*np.sqrt(p1*q1/n+p2*q2/n)
lcl, ucl

(-0.011318564551246066, 0.051318564551246094)

# 문제

- H0 : A, B 회사 사이의 타이어 제동거리 간 차이는 없다.
- H1 : A, B 회사 사이의 타이어 제동거리 간 차이는 있다.

In [100]:
mean_A=13.5
mean_B=14.1
var_A=1.46**2
var_B=1.33**2
n_A=64
n_B=36

In [101]:
rv=stats.norm()
z=(mean_A-mean_B)/np.sqrt(var_A/n_A+var_B/n_B)
z

-2.0896619735236412

In [102]:
p=rv.cdf(z)*2
p

0.03664817519569912

귀무가설을 기각한다. 즉 A,B 회사 사이의 평균 타이어 제동거리 간 차이는 있다.

# 문제

- H0 : A, B 지역 쌀 생산량에는 차이가 없다.
- H1 : A, B 지역 쌀 생산량에는 차이가 있다

In [114]:
mean_A=364
mean_B=330
n_A=15
n_B=16
var_A=45**2
var_B=60**2

In [115]:
sp=np.sqrt(((n_A-1)*var_A+(n_B-1)*var_B)/(n_A+n_B-2))
sp

53.28841499250839

In [119]:
t=(mean_A-mean_B)/(sp*np.sqrt(1/n_A+1/n_B))
t

1.775296464501227

In [112]:
rv=stats.t(df=n_A+n_B-2)
t

0.22931091246210092

In [113]:
alpha=0.05
rv.interval(1-alpha)

(-2.045229642132703, 2.045229642132703)

검정통계량이 기각역 안에 놓이지 않으므로 귀무가설을 기각할 수 없다. 즉 A, B지역 쌀 생산량에는 차이가 없다.

- H0 : 기존의 약과 새로개발한 신약 혈압사이에는 차이가 없다
- H1 : 새로 개발한 약이 혈압이 더 낮다(less)

In [121]:
df=pd.DataFrame({'기존':[157, 182, 183, 192, 165, 196], '신약':[155, 185, 181, 188, 167, 183]})
df

기존   신약
0  157  155
1  182  185
2  183  181
3  192  188
4  165  167
5  196  183

In [122]:
stats.ttest_rel(df['기존'], df['신약'], alternative='less')

Ttest_relResult(statistic=1.142857142857143, pvalue=0.8475742939802655)

- 문제

- H0 : 우리나라의 음주율은 20.5%이다.
- H1 : 우리나라의 음주율은 20.5%가 아니다.

In [137]:
n=500
p=0.205
q=1-p
p_r=120/500
z=(p_r-p)/np.sqrt(p*q/n)
z

1.9386193875077975

In [136]:
alpha=0.05
rv.interval(1-alpha)

(-1.959963984540054, 1.959963984540054)

기각역안에 들지 않으므로 귀무가설을 기각할 수 없다. 즉 우리나라의 음주율은 20.5%이다.

In [133]:
np.sqrt(p*q/n)

9.093949042870822e-79

In [134]:
p

4.236675675939096e-154

- H0 : 청년층과 장년층 사이에는 차이가 없다.
- H1 : 청년층과 장년층 사이에는 차이가 있다.

In [ ]:
p